modules: mbedtls: add mbed debug dependency to its Kconfig

emillindq opened the pull request:
Debug function pointer in MbedTLS will not be set if not
sockets log level is debug (see [sockets_tls:412)

https://github.com/zephyrproject-rtos/zephyr/blob/8e75fff5808f0af0e8de941f6a5d77de72962b0b/subsys/net/lib/sockets/sockets_tls.c#L411-L413

rlubos reviewed (commented):
I don't think this is the right approach - mbedTLS is a standalone module and should not be dependent on the socket layer.

emillindq commented:
> I don't think this is the right approach - mbedTLS is a standalone module and should not be dependent on the socket layer.

The module is not dependent on it; only its debug prints which is actually the truth since it's some LOG_DBG call in sockets_tls that prints the messages from mbedTLS. No LOG_DBG in sockets_tls, no debug msg from mbedTLS.

As it is now, people will enable `CONFIG_MBEDTLS_DEBUG` and expect mbedTLS debug prints to come out but nothing happens (which is what just happened to me). At least now they will get an error and be able to fix it.

mniestroj commented:
Maybe let's move `tls_init` outside of `socket_tls.c`, remove `#if (CONFIG_NET_SOCKETS_LOG_LEVEL >= LOG_LEVEL_DBG)` part and hardcode `LOG_LEVEL_DBG` whenever `CONFIG_MBEDTLS_DEBUG` is selected? At some point we add log level mapping between mbedTLS and Zephyr's logging, so that mbedTLS logs would be more readable (especially errors and warnings).

rlubos commented:
> The module is not dependent on it; only its debug prints which is actually the truth since it's some LOG_DBG call in sockets_tls that prints the messages from mbedTLS. No LOG_DBG in sockets_tls, no debug msg from mbedTLS.

This is exactly what I understand as adding sockets dependency to the mbedTLS - what if someone wanted to use mbedTLS w/o socket layer and enable debug logs in it (and provide his own debug handler, note that the debug log function is registered per TLS context). It won't be possible with such approach.

> As it is now, people will enable CONFIG_MBEDTLS_DEBUG and expect mbedTLS debug prints to come out but nothing happens (which is what just happened to me). At least now they will get an error and be able to fix it.

If that's an issue then I think it should be solved in a different way. At very least, we could bump the default socket log level to DEBUG if `MBEDTLS_DEBUG` is enabled. But I think that a better solution would be decoupling the mbedTLS logs from socket layer logs, i. e. register a separate LOG instance for mbedTLS with its own log level (there already is a log level specified for mbedTLS), and then register the log handler in the TLS context independently from `CONFIG_NET_SOCKETS_LOG_LEVEL`.

emillindq commented:
> > The module is not dependent on it; only its debug prints which is actually the truth since it's some LOG_DBG call in sockets_tls that prints the messages from mbedTLS. No LOG_DBG in sockets_tls, no debug msg from mbedTLS.
> 
> This is exactly what I understand as adding sockets dependency to the mbedTLS - what if someone wanted to use mbedTLS w/o socket layer and enable debug logs in it (and provide his own debug handler, note that the debug log function is registered per TLS context). It won't be possible with such approach.
> 
> > As it is now, people will enable CONFIG_MBEDTLS_DEBUG and expect mbedTLS debug prints to come out but nothing happens (which is what just happened to me). At least now they will get an error and be able to fix it.
> 
> If that's an issue then I think it should be solved in a different way. At very least, we could bump the default socket log level to DEBUG if `MBEDTLS_DEBUG` is enabled. But I think that a better solution would be decoupling the mbedTLS logs from socket layer logs, i. e. register a separate LOG instance for mbedTLS with its own log level (there already is a log level specified for mbedTLS), and then register the log handler in the TLS context independently from `CONFIG_NET_SOCKETS_LOG_LEVEL`.

Ah yes I understand, you are right. We can't use mbedTLS without socket layer as it is now. I agree with you, we need to decoulple it. I'll look at it this weekend.

jukkar commented:
> register a separate LOG instance for mbedTLS with its own log level (there already is a log level specified for mbedTLS), and then register the log handler in the TLS context independently from `CONFIG_NET_SOCKETS_LOG_LEVEL`

Yes, this is actually the best option here.

emillindq committed:
modules: mbedtls: move mbedTLS logginq to separate module

The goal of this merge request is to move mbedTLS debug
prints to a separate log module, so as to not be dependent
on sockets log module which would require additional
non-obvious config enabled in order to see mbedTLS debug
prints.

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

emillindq commented:
Please see updated branch.

If I build echo_client sample with 
```
west build -b stm32h747i_disco_m7 zephyr/samples/net/sockets/echo_client/ -- -DCONF_FILE="prj.conf overlay-tls.conf"
```
and the prj.conf from below (and no echo server to provoke errors and warnings) I can see that mbedtls debug messages comes out. I have also verified log level filtering works (as in, if I set `CONFIG_MBEDTLS_DEBUG_LEVEL=3` we will only see error, warning & info).

Please note that:
1. I have kept `mbedtls_debug_set_threshold(CONFIG_MBEDTLS_DEBUG_LEVEL);` in `sockets_tls.c` since this is a global setting among all mbedtls contexts.
2. I changed the default `CONFIG_MBEDTLS_DEBUG_LEVEL` to 3 since as a user I'd expect some kind of logs to come out if I enable `CONFIG_MBEDTLS_DEBUG` (which is default n so no debug messages will come out anyways if you don't enable this explicitly).
```
# Generic networking options
CONFIG_NETWORKING=y
CONFIG_NET_UDP=y
#CONFIG_NET_TCP=y
#CONFIG_NET_IPV6=y
CONFIG_NET_IPV4=y
CONFIG_NET_SOCKETS=y
CONFIG_NET_SOCKETS_POSIX_NAMES=y
CONFIG_NET_SOCKETS_POLL_MAX=4
CONFIG_NET_CONNECTION_MANAGER=y

# Kernel options
CONFIG_MAIN_STACK_SIZE=2048
CONFIG_ENTROPY_GENERATOR=y
CONFIG_TEST_RANDOM_GENERATOR=y
CONFIG_INIT_STACKS=y

# Logging
CONFIG_NET_LOG=y
CONFIG_LOG=y
CONFIG_NET_STATISTICS=y
CONFIG_PRINTK=y

# Network buffers
CONFIG_NET_PKT_RX_COUNT=16
CONFIG_NET_PKT_TX_COUNT=16
CONFIG_NET_BUF_RX_COUNT=80
CONFIG_NET_BUF_TX_COUNT=80
CONFIG_NET_CONTEXT_NET_PKT_POOL=y

# IP address options
#CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=3
#CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=4
CONFIG_NET_MAX_CONTEXTS=10

# Network shell
CONFIG_NET_SHELL=y

# The addresses are selected so that qemu<->qemu connectivity works ok.
# For linux<->qemu connectivity, create a new conf file and swap the
# addresses (so that peer address is ending to 2).
CONFIG_NET_CONFIG_SETTINGS=y
#CONFIG_NET_CONFIG_NEED_IPV6=y
#CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::2"
#CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::1"
CONFIG_NET_CONFIG_NEED_IPV4=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.2"
CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.1"
CONFIG_MBEDTLS_DEBUG=y
CONFIG_LOG_STRDUP_BUF_COUNT=100
```

jukkar commented on subsys/net/lib/sockets/mbedtls_dbg.c:
```diff
@@ -0,0 +1,55 @@
+/*
```
As there is nothing network specific any more in `mbedtls_dbg.[c|h]`, we could place the files in some more generic place. One such place could be `lib/os`, what do you think?

rlubos replied:
I was about to suggest the same, it seems that module-specific code could also be places in `modules/mbedtls/`.

jukkar replied:
> module-specific code could also be places in `modules/mbedtls/`

+1 to that

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -309,7 +288,7 @@ static int tls_init(const struct device *unused)
 	}
 
-#if defined(MBEDTLS_DEBUG_C) && (CONFIG_NET_SOCKETS_LOG_LEVEL >= LOG_LEVEL_DBG)
+#if defined(MBEDTLS_DEBUG_C) && (CONFIG_MBEDTLS_DEBUG)
```
The `dbg_init()` and relevant setup calls should be called somewhere more generic place, one option is `kernel/init.c` just before `main()` is called.

rlubos replied:
I'm afraid that won't work, `mbedtls_dbg_init()` and thus `mbedtls_ssl_conf_dbg()` requires a TLS context structure, and you get this only after socket creation.

rlubos replied:
Another thing is, there already is a mbedTLS initialized, but it's placed in the `mbedtls` tree: https://github.com/zephyrproject-rtos/mbedtls/blob/master/zephyr_init.c. I wonder if this should rather be moved into the Zephyr tree since it's a glue code for Zephyr. We could configure the threshold in there then. But it's probably out of the scope of this PR.

jukkar replied:
> I wonder if this should rather be moved into the Zephyr tree since it's a glue code for Zephyr. We could configure the threshold in there then. But it's probably out of the scope of this PR.

Sounds like a plan.

emillindq replied:
@rlubos Please correct me if I'm wrong, but you suggest moving mbedtls_dbg.c file to modules/crypto/mbedtls/mbedtls_dbg.c and then keep the call in sockets_tls.c, and then move the setting of logging threshold to zephyr_init?

Just asking for clarification to reduce unnecessary wrong commits.

I'm aiming for this way since it seems cleaner and also since I kinda have a hard time understanding how CMake includes source files in zephyr/modules/* :smile:

rlubos replied:
> Please correct me if I'm wrong, but you suggest moving mbedtls_dbg.c file to modules/crypto/mbedtls/mbedtls_dbg.c and then keep the call in sockets_tls.c, and then move the setting of logging threshold to zephyr_init?

Correct. As for the `zephyr_init`, I'd consider moving it into the Zephyr three from the mbedTLS module, but I understand if you don't want to touch it as it's a bit out of scope of this PR. In such case, I guess we could leave the threshold setting as it is for now.


> I'm aiming for this way since it seems cleaner and also since I kinda have a hard time understanding how CMake includes source files in zephyr/modules/*
hal_nordic already does it, you can use it as an example: https://github.com/zephyrproject-rtos/zephyr/tree/master/modules/hal_nordic. I guess you just need to add top-level CMakeLists.txt in the module top folder which will add the source files.

mniestroj commented on subsys/net/lib/sockets/mbedtls_dbg.c:
```diff
@@ -0,0 +1,55 @@
+		}
+	}
+
+#define LOG_ARGS "%s:%04d: %s", basename, line, log_strdup(str)
+	switch (level) {
+	case 1:
+		LOG_ERR(LOG_ARGS);
+		break;
+	case 2:
+		LOG_WRN(LOG_ARGS);
+		break;
+	case 3:
+		LOG_INF(LOG_ARGS);
+		break;
+	case 4:
+		LOG_DBG(LOG_ARGS);
+		break;
+	default:
+		LOG_ERR("mbedTLS invalid debug level (%d)", level);
+		break;
```
I don't like this `LOG_ARGS` macros. What about using `Z_LOG()` and pass log level there? It is what `lvgl` glue code already uses.

jukkar replied:
Why should we use some internal macro `Z_LOG()` here? The `LOG_xxx()` macro will call `Z_LOG()` anyway but we should not use those internal macros meant for logging subsystem.

mniestroj replied:
True, `z_` prefix denotes internal API. Using `Z_LOG()` allows to reduce the code, which is why I suggested that. Breaking `Z_LOG()` API can still be easily followed by updating this file, as long it is in the same repo. Each approach has of course pros and cons. Personally I would prefer to use `Z_LOG()` here, but this is just a slight preference, not a requirement.

jukkar replied:
I think we should stick with the public `LOG_xxx` macros, the level variable maps conveniently to zephyr log levels so it could be used directly, but still I would use the public ones here.

emillindq committed:
tmp

pfalcon reviewed (commented):
Some additional suggestions. 

@emillindq, what's the status of this PR?

pfalcon commented:
Additionally, the commit message has typo:

> modules: mbedtls: move mbedTLS loggin**q** to separate module

emillindq commented:
> Some additional suggestions.
> 
> @emillindq, what's the status of this PR?

The status is that I'm hoping to get time to complete it very soon (will try this weekend). The plan is to implement https://github.com/zephyrproject-rtos/zephyr/pull/32473#discussion_r592994470 and I've gotten quite a way already. If anybody wants to finish it, please do.

pfalcon commented:
@emillindq 

> The status is that I'm hoping to get time to complete it very soon (will try this weekend). The plan is to implement #32473 (comment) and I've gotten quite a way already. If anybody wants to finish it, please do.

That's good plan. Otherwise, just a friendly ping that 2.6 pre-release freeze in coming up (1-2 weeks, not sure), during which it may be harder to merge changes (e.g. may need to wait for reopening merge window for the next release).

emillindq commented:
Unfortunately I'm not sure when I'm going to be able to complete this. I encourage somebody else to continue where I left off, but in the meantime I'm closing this as this is a non-critical thing as long as you know to enable the right configs.
